src/cpu/aarch64/acl_pooling.cpp

```diff
@@ -21,33 +21,265 @@ namespace impl {
+}
+
+// Generally, ACL is faster above a per thread problem size 'cutoff'.
+// The value of this cutoff has been found empirically,
+// through profiling on a Neoverse-N1 cpu.
+// Note: This rule is approximate, Not all problems follow this rule.
+//
+// Parameters used in the heuristics:
+// - problem_size: defined as mb * ic * oh * ow * kh * kw
+// - thread_count
+// - is_nhwc (as opposed to nchw)
+// - use_square_acl_kernels: For nchw pooling, acl has faster kernels
+//   for pooling window (kernel) sizes of 2x2, 3x3, or 7x7.
+// - is_training: Max pooling training cases require a workspace tensor,
+//   so these cases have been implemented in a seperate kernel in ACL.
+bool acl_pooling_fwd_t::pd_t::use_acl_avg_pool_heuristic(int problem_size,
+        int thread_count, bool is_nhwc, bool use_square_acl_kernel) {
+    int cutoff;
+    if (is_nhwc) {
+        if (thread_count == 1)
+            cutoff = 200;
+        else
+            cutoff = 4096;
+    } else {
+        if (use_square_acl_kernel) {
+            if (thread_count == 1)
+                cutoff = 100;
+            else if (thread_count > 32)
+                return false;
+            else
+                cutoff = 2048;
+        } else
+            return false;
+    }
+    return problem_size > cutoff * thread_count;
+}
+
+bool acl_pooling_fwd_t::pd_t::use_acl_max_pool_heuristic(int problem_size,
+        int thread_count, bool is_nhwc, bool use_square_acl_kernel,
+        bool is_training) {
+    int cutoff;
+    if (is_nhwc) {
+        if (thread_count == 1)
+            cutoff = 200;
+        else {
+            if (is_training)
+                cutoff = 2048;
+            else
+                cutoff = 4096;
+        }
+    } else {
+        if (use_square_acl_kernel) {
+            if (thread_count == 1)
+                cutoff = 100;
+            else
+                cutoff = 1024;
+        } else {
+            if (thread_count == 1)
+                return true;
+            else if (thread_count > 16)
+                return false;
+            else
+                cutoff = 25000;
+        }
+    }
+    return problem_size > cutoff * thread_count;
+}
+
+status_t acl_pooling_fwd_t::pd_t::init_scratchpad(
+        memory_tracking::registrar_t &scratchpad) {
+    const memory_desc_wrapper dst_d(&dst_md_);
+    scratchpad.book(memory_tracking::names::key_pool_reduction, dst_d.nelems(),
+            sizeof(float));
+    if (asp_.use_ws) {
+        scratchpad.book(memory_tracking::names::key_pool_ind_plain2blocked_cvt,
+                dst_d.nelems(), sizeof(uint32_t));
+    }
+    return status::success;
+}
+
+acl_pooling_fwd_t::acl_pooling_fwd_t(const pd_t *apd)
+    : primitive_t(apd)
+    , pooling_op_ {
+              std::make_unique<arm_compute::experimental::op::CpuPool2d>()} {}
+
+status_t acl_pooling_fwd_t::execute(const exec_ctx_t &ctx) const {
+    return execute_forward(ctx);
+}
+
+const acl_pooling_fwd_t::pd_t *acl_pooling_fwd_t::pd() const {
```

Comment:
Thanks for moving the definitions to the `.cpp` file. However, I should have mentioned this particular function might be better left in the `.hpp` file [as discussed here](https://github.com/uxlfoundation/oneDNN/pull/2992#discussion_r2153657576). We can probably mark it `inline` as well, as done in that patch.